File: ex95.java
// Devide and Conquer

import java.util.List;
import java.util.ArrayList;
import java.util.LinkedList;

// Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) { val = x; }
}


class Solution {
    public static List<TreeNode> DAC(int lo, int hi) {
        List<TreeNode> result = new ArrayList<TreeNode>();
        
        // Opt out of recursion
        if (lo > hi) {
            result.add(null);
            return result;
        }
        
        for (int i = lo; i <= hi; i++) {
            // Divide (Recursion)
            List<TreeNode> leftTree = DAC(lo, i - 1);
            List<TreeNode> rightTree = DAC(i + 1, hi);

            // conquer
            for (TreeNode l : leftTree) {
                for (TreeNode r : rightTree) {
                    TreeNode root = new TreeNode(i);
                    root.left = l;
                    root.right = r;
                    
                    result.add(root);
                }
            }
        }      
        return result;
    }
    
    
    public List<TreeNode> generateTrees(int n) {
        if (n == 0)    return new LinkedList<TreeNode>();
        
        List<TreeNode> root = DAC(1, n);
        
        return root;
    }
}

File: code/ex95.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> generateTrees(int n) {
        vector<TreeNode*> res;
        vector<vector<vector<TreeNode*>>> dp(n + 2, vector<vector<TreeNode*>>(n + 2, vector<TreeNode*>()));
        for (int i = 1; i <= n; i++)    dp[i][i].push_back(new TreeNode(i));
        
        for (int l = 2; l <= n; l++) {
            for (int i = 1; i <= n + 1 - l; i++) {
                int j = i + l - 1;
                
                for (int mid = i; mid <= j; mid++) {
                    vector<TreeNode*> leftNodes = dp[i][mid - 1];
                    vector<TreeNode*> rightNodes = dp[mid + 1][j];

                    if (leftNodes.size() == 0) {
                        for (TreeNode* right : rightNodes) {
                            dp[i][j].push_back(generateNode(mid, NULL, right));
                        }
                    }

                    else if (rightNodes.size() == 0) {
                        for (TreeNode* left : leftNodes) {
                            dp[i][j].push_back(generateNode(mid, left, NULL));
                        }
                    }

                    else {
                        for (TreeNode* left : leftNodes) {
                            for (TreeNode* right : rightNodes) {
                                dp[i][j].push_back(generateNode(mid, left, right));
                            }
                        }
                    }
                }
            }
        }
        
        return dp[1][n];
    }
    
private:
    TreeNode* generateNode(int rootVal, TreeNode* left, TreeNode* right) {
        TreeNode* root = new TreeNode(rootVal);
        root->left = left;
        root->right = right;
        return root;
    }
};